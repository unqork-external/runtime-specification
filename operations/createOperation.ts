import type { OperationTypes } from './enums/operation-types.enum'
import { OperationTypeMap } from './typemaps/operation.typemap'
import type { CustomOperationTypeMap } from '../../operations'
import type { InputRef } from '../inputs/inputRef'

/**
 * Given an OperationType, Typescript will complete the options list.
 *
 * In return, the caller receives Typescript constructed class.
 *
 * This includes preset defaults.
 *
 * In essence, this function ensures that all keys are fully present, fully typed, and ready for runtime
 * validation if necessary.
 *
 * @param type
 * @param options
 * @param name name of the operation
 * @param additional
 * @param spreadOp
 * @param additionalArgs
 * @returns - The Operation Class type
 */
export const createOperation = <OpType extends OperationTypes>(
  type: OpType,
  options: InstanceType<(typeof OperationTypeMap)[OpType]>['options'],
  additionalArgs?: Partial<AdditionalArgs>,
): InstanceType<(typeof OperationTypeMap)[OpType]> => {
  const classOpType = OperationTypeMap[type]
  const op = new classOpType()
  return Object.assign(op, {
    type,
    options: {
      ...op.options,
      ...options,
    },
    ...additionalArgs,
  })
}

export const createCustomOperation = <OperationEnum extends string, OpType extends OperationEnum>(
  type: OpType,
  options: InstanceType<CustomOperationTypeMap<OperationEnum>[OpType]['model']>['options'],
  typeMap: CustomOperationTypeMap<OperationEnum>,
  additionalArgs?: Partial<AdditionalArgs>,
) => {
  const classOpType = typeMap[type].model
  const op = new classOpType()
  return Object.assign(op, {
    type,
    options: {
      ...op.options,
      ...options,
    },
    ...additionalArgs,
  })
}

interface AdditionalArgs {
  name: string
  notifyImmediately: boolean
  creatorSummary: string
}
