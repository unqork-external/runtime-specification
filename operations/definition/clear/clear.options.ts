import { Description, Example, Required } from '@tsed/schema'

import { OperationOptions } from '../../interface'

export class ClearOptions extends OperationOptions {
  @Required()
  @Example('textfield')
  @Description('key of target to apply the clear operation')
  public targetKey: string

  @Example(['textfield', 'textarea'])
  @Description('List of component types to ignore when the clear operation runs.')
  DO_NOT_USE_ignoreChildComponentTypes?: string[]
}
