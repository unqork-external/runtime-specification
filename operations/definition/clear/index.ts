export * from './clear.options'
export * from './clear.model'
