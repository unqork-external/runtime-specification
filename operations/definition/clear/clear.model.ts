import { DiscriminatorValue, Required } from '@tsed/schema'

import { ClearOptions } from './clear.options'
import { Alpha } from '../../../../decorators'
import { OperationTypes } from '../../enums/operation-types.enum'
import { Operation } from '../../interface/operations.interface'

@DiscriminatorValue(OperationTypes.CLEAR)
@Alpha()
export class ClearOperation extends Operation<OperationTypes.CLEAR> {
  public type = OperationTypes.CLEAR as const

  @Required()
  public options: ClearOptions = new ClearOptions()
}
