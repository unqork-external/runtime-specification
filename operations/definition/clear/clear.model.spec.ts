import { ClearOperation } from './clear.model'
import { generateSchemaAndValidate } from '../../../../utilities'
import { createOperation } from '../../createOperation'
import { OperationTypes } from '../../enums/operation-types.enum'

describe('Clear Operation Model Test', () => {
  let validate, schema
  beforeAll(() => {
    const response = generateSchemaAndValidate(ClearOperation)
    validate = response.validate
    schema = response.schema
  })

  it('should match this JSON Schema', () => {
    expect(schema).toStrictEqual({
      definitions: {
        ClearOptions: {
          properties: {
            DO_NOT_USE_ignoreChildComponentTypes: {
              description: 'List of component types to ignore when the clear operation runs.',
              examples: [['textfield', 'textarea']],
              type: 'array',
            },
            targetKey: {
              description: 'key of target to apply the clear operation',
              examples: ['textfield'],
              minLength: 1,
              type: 'string',
            },
          },
          required: ['targetKey'],
          type: 'object',
        },
      },
      properties: {
        creatorSummary: {
          description: 'A detailed summary of the operation',
          type: 'string',
        },
        options: {
          $ref: '#/definitions/ClearOptions',
        },
        type: {
          const: 'CLEAR',
          examples: ['CLEAR'],
          type: 'string',
        },
        name: {
          description: 'Name of the operation',
          type: 'string',
        },
      },
      required: ['options'],
      type: 'object',
    })
  })

  it('should validate appropriately', () => {
    const op = createOperation(OperationTypes.CLEAR, { targetKey: 'abc-123' })
    const isValid = validate(op)
    expect(isValid).toBeTrue()
  })

  it('should set up the clear operation properly', () => {
    const op = createOperation(OperationTypes.CLEAR, { targetKey: 'abc-123' })
    expect(op.options.targetKey).toEqual('abc-123')
  })
})
