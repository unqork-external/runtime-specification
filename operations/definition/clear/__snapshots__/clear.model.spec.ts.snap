// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clear Operation Model Test should match this JSON Schema 1`] = `
{
  "definitions": {
    "ClearOptions": {
      "properties": {
        "DO_NOT_USE_ignoreChildComponentTypes": {
          "description": "List of component types to ignore when the clear operation runs.",
          "examples": [
            [
              "textfield",
              "textarea",
            ],
          ],
          "type": "array",
        },
        "shouldExecute": {
          "default": true,
          "description": "
    Conditionally execute this operation.
    Can set to false to temporarily disabled when debugging.
  ",
          "type": [
            "string",
            "boolean",
          ],
        },
        "targetKey": {
          "description": "key of target to apply the clear operation",
          "examples": [
            "textfield",
          ],
          "minLength": 1,
          "type": "string",
        },
      },
      "required": [
        "targetKey",
      ],
      "type": "object",
    },
  },
  "properties": {
    "creatorSummary": {
      "description": "A detailed summary of the operation",
      "type": "string",
    },
    "name": {
      "description": "Name of the operation",
      "type": "string",
    },
    "options": {
      "$ref": "#/definitions/ClearOptions",
    },
    "type": {
      "const": "CLEAR",
      "examples": [
        "CLEAR",
      ],
      "type": "string",
    },
  },
  "required": [
    "options",
  ],
  "type": "object",
}
`;
